[GTK][WPE] Make Damage iterable

carlosgcampos opened the pull request:
#### 63ebc0af1d9d8bb676eced504beba71393363392
<pre>
[GTK][WPE] Make Damage iterable
<a href="https://bugs.webkit.org/show_bug.cgi?id=291358">https://bugs.webkit.org/show_bug.cgi?id=291358</a>

Reviewed by Alejandro G. Castro.

Instead of exposing the internal vector we use for dirty rects, that can
contain empty rectangles after uniting, we can make Damage iterable and
always return the actual rectangles.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::size const):
(WebCore::Damage::isEmpty const):
(WebCore::Damage::bounds const):
(WebCore::Damage::iterator::iterator):
(WebCore::Damage::iterator::operator*):
(WebCore::Damage::iterator::operator++):
(WebCore::Damage::iterator::operator== const):
(WebCore::Damage::begin const):
(WebCore::Damage::end const):
(WebCore::Damage::copyToVector const):
(WebCore::Damage::rectsForPainting const):
(WebCore::Damage::makeFull):
(WebCore::Damage::add):
(WebCore::Damage::initialize):
(WebCore::Damage::uniteExistingRects):
(WebCore::Damage::cellIndexForRect const):
(WebCore::Damage::unite):
(WebCore::FrameDamageHistory::addDamage):
(WebCore::operator&lt;&lt;):
(WebCore::Damage::rects const): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
(WebCore::TextureMapperDamageVisualizer::paintDamage const):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::collectDamageSelf):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStore::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::setDirtyRegion):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, Basics)):
(TestWebKitAPI::TEST(Damage, Mode)):
(TestWebKitAPI::TEST(Damage, Move)):
(TestWebKitAPI::TEST(Damage, AddRect)):
(TestWebKitAPI::TEST(Damage, AddRects)):
(TestWebKitAPI::TEST(Damage, AddDamage)):
(TestWebKitAPI::TEST(Damage, Unite)):
(TestWebKitAPI::TEST(Damage, RectsForPainting)):
(TestWebKitAPI::TEST(Damage, MaxRectangles)):

Canonical link: <a href="https://commits.webkit.org/294236@main">https://commits.webkit.org/294236@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3ff4311a8ee8e4c1cfe84cf9d4e455153a2ad01

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100290 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10237 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105420 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50873 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102331 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20244 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28413 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76364 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33420 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103297 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15494 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90589 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56720 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15315 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8588 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50241 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85227 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8670 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107779 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27403 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20095 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85316 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27766 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86796 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84854 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29533 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7272 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21309 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16463 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27340 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32573 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27151 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30467 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28710 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -106,25 +108,139 @@ class Damage {
                 IntRect minimumBoundingRectangleContaingOverlaps;
-                for (const auto& rect : m_rects) {
+                for (const auto& rect : m_rects.rects) {
                     if (!rect.isEmpty())
```
I think we can use the iterator here and avoid the `isEmpty` check

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -254,13 +375,41 @@ class Damage {
+
+            // When damage is empty and we know we will need to unite, we can unite the rects directly.
+            if (isEmpty()) {
+                auto rectsCount = other.size();
```
could be `const` or removed

Scony commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -311,46 +459,50 @@ class Damage {
-            m_rects[0] = m_minimumBoundingRectangle;
+        if (m_rects.rects.size() == 1) {
+            m_rects.rects[0] = m_minimumBoundingRectangle;
+            m_rects.indices.set(0);
```
in this case we have a guarantee that all `m_rects` are non-empty so perhaps we could clear the `m_rects.indices` thus having a faster codepath in `at()`?

carlosgcampos replied:
But we don't know if more rects will be added later, when the second one is added we would be missing the first index. We could handle vector size == 1 size in at() and return always the first element in that case. Or we could add the first index to the BitVector when the second element is added.

Scony replied:
This is `unite` function we're talking about, so if we have `m_rects.rects.size() == 1` it means that we have a 1x1 grid and the rects cannot grow further, no?

carlosgcampos replied:
You are right, once should unite is true the vector can't grow.

Scony reviewed (commented):
LGTM - I'll do the testing with it tomorrow

alexgcastro reviewed (approved):
LGTM!

carlosgcampos committed:
[GTK][WPE] Make Damage iterable
https://bugs.webkit.org/show_bug.cgi?id=291358

Reviewed by Alejandro G. Castro.

Instead of exposing the internal vector we use for dirty rects, that can
contain empty rectangles after uniting, we can make Damage iterable and
always return the actual rectangles.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::size const):
(WebCore::Damage::isEmpty const):
(WebCore::Damage::bounds const):
(WebCore::Damage::iterator::iterator):
(WebCore::Damage::iterator::operator*):
(WebCore::Damage::iterator::operator++):
(WebCore::Damage::iterator::operator== const):
(WebCore::Damage::begin const):
(WebCore::Damage::end const):
(WebCore::Damage::copyToVector const):
(WebCore::Damage::rectsForPainting const):
(WebCore::Damage::makeFull):
(WebCore::Damage::add):
(WebCore::Damage::initialize):
(WebCore::Damage::uniteExistingRects):
(WebCore::Damage::cellIndexForRect const):
(WebCore::Damage::unite):
(WebCore::FrameDamageHistory::addDamage):
(WebCore::operator<<):
(WebCore::Damage::rects const): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
(WebCore::TextureMapperDamageVisualizer::paintDamage const):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::collectDamageSelf):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStore::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::setDirtyRegion):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, Basics)):
(TestWebKitAPI::TEST(Damage, Mode)):
(TestWebKitAPI::TEST(Damage, Move)):
(TestWebKitAPI::TEST(Damage, AddRect)):
(TestWebKitAPI::TEST(Damage, AddRects)):
(TestWebKitAPI::TEST(Damage, AddDamage)):
(TestWebKitAPI::TEST(Damage, Unite)):
(TestWebKitAPI::TEST(Damage, RectsForPainting)):
(TestWebKitAPI::TEST(Damage, MaxRectangles)):

Canonical link: https://commits.webkit.org/294236@main
